### PR TITLE
Add qgsgeometry_cast

### PR DESCRIPTION
**Pull request by m-kuhn**
Because it
- is faster than dynamic_cast
- doesn't rely on RTTI
- encapsulates type checking in one place

### CODE REVIEW

File: src/core/geometry/qgsabstractgeometry.h [hunk 1]
```diff
@@ -484,4 +484,12 @@ struct CORE_EXPORT QgsVertexId
   VertexType type;
 };
 
+#ifndef SIP_RUN
+template <class T>
+inline T qgsgeometry_cast( const QgsAbstractGeometry *geom )
+{
+  return const_cast<T>( reinterpret_cast<T>( 0 )->cast( geom ) );
```
Comment (nyalldawson) [inline src/core/geometry/qgsabstractgeometry.h]: shouldn't this be static_cast?

Comment (m-kuhn) [inline src/core/geometry/qgsabstractgeometry.h]: Which one?
I took `qobject_cast` as a reference but not completely sure I got everything right.

Comment (nyalldawson) [inline src/core/geometry/qgsabstractgeometry.h]: Instead of the reinterpret_cast. static_cast should be safer (assuming it compiles ok)

Comment (m-kuhn) [inline src/core/geometry/qgsabstractgeometry.h]: I think there's not much to be typechecked or safe at any time with casting a `nullptr` I guess :P

Comment (nyalldawson) [inline src/core/geometry/qgsabstractgeometry.h]: Oh ... I didn't even see that bit :p that'll teach me for skim-reviewing

**Comment by nyalldawson** [comment]
@m-kuhn any change of reviving this? I can see benefits here....

**Pull request merged**